File: rust/crates/corlinman-gateway/src/grpc/placeholder.rs
//! gRPC wrapper over [`corlinman_core::placeholder::PlaceholderEngine`].
//!
//! Direction: Python client → Rust server (reverse of agent.proto). The
//! Python `context_assembler` dials this service on the UDS path in
//! `$CORLINMAN_UDS_PATH` (default `/tmp/corlinman.sock`) and calls
//! `Render` for every template it wants expanded before a provider call.
//!
//! Scope for B1-BE3 (BRIDGE): this is the plumbing only. The engine
//! registered here has no resolvers yet — real namespace resolvers
//! (skills, variables, agent cards, …) land in Batch 2 (B2-BE4). Tokens
//! with a namespace that has no resolver round-trip back unchanged and
//! are surfaced in [`RenderResponse::unresolved_keys`] for observability.
//!
//! Error mapping preserves the enum shape from
//! [`corlinman_core::placeholder::PlaceholderError`]:
//!
//! | engine error            | `error` string          |
//! |-------------------------|-------------------------|
//! | `Cycle(k)`              | `"cycle:<k>"`           |
//! | `DepthExceeded {..}`    | `"depth_exceeded"`      |
//! | `Resolver {..}`         | `"resolver:<msg>"`      |

use std::path::{Path, PathBuf};
use std::sync::Arc;

use corlinman_core::placeholder::{DynamicResolver, PlaceholderCtx, PlaceholderEngine};
use corlinman_core::CorlinmanError;
use corlinman_memory_host::MemoryHost;
use corlinman_proto::v1::{
    placeholder_server::{Placeholder, PlaceholderServer},
    RenderRequest, RenderResponse,
};
use once_cell::sync::Lazy;
use regex::Regex;
use tokio::net::UnixListener;
use tokio_stream::wrappers::UnixListenerStream;
use tonic::{transport::Server, Request, Response, Status};

/// Default UDS path the Rust side binds for Python→Rust traffic. Kept
/// separate from `/tmp/corlinman-py.sock` so the two sides can be
/// restarted independently without stepping on each other's socket file.
pub const DEFAULT_RUST_SOCKET: &str = "/tmp/corlinman.sock";

/// Env var the Python client honours (and the server respects when set).
/// Matches the contract documented on `PlaceholderClient`.
pub const ENV_RUST_SOCKET: &str = "CORLINMAN_UDS_PATH";

/// Same regex as the engine, used here to harvest unresolved token keys
/// from the post-render output. Kept local because `corlinman-core` does
/// not publicly export its compiled regex.
static TOKEN_RE: Lazy<Regex> = Lazy::new(|| {
    Regex::new(r"\{\{([^{}]*?)\}\}").expect("placeholder regex is a compile-time constant")
});

/// gRPC service shell. Wraps a shared [`PlaceholderEngine`] so multiple
/// concurrent `Render` RPCs share the same resolver registry.
pub struct PlaceholderService {
    engine: Arc<PlaceholderEngine>,
}

impl PlaceholderService {
    /// Build a service around an already-populated engine. Production boot
    /// constructs the engine once (with every resolver registered) and
    /// threads the same `Arc` here.
    pub fn new(engine: Arc<PlaceholderEngine>) -> Self {
        Self { engine }
    }

    /// Convenience for tests + the B1-BE3 bridge boot: stand up a service
    /// backed by a brand-new empty engine (no resolvers). Every token
    /// whose namespace is not a registered static key will round-trip
    /// back through `unresolved_keys`.
    pub fn with_empty_engine() -> Self {
        Self::new(Arc::new(PlaceholderEngine::new()))
    }
}

/// Build the production placeholder engine.
///
/// The gateway owns the resolver registrations because both namespaces need
/// gateway-local state: `episodes` reads per-tenant SQLite files under the
/// data dir, while `memory` queries the same [`MemoryHost`] surfaced over MCP
/// and `/memory/*`.
pub fn build_engine(
    data_dir: impl Into<PathBuf>,
    memory_host: Option<Arc<dyn MemoryHost>>,
) -> PlaceholderEngine {
    build_engine_with_episodes(data_dir, memory_host, None)
}

fn build_engine_with_episodes(
    data_dir: impl Into<PathBuf>,
    memory_host: Option<Arc<dyn MemoryHost>>,
    episodes: Option<Arc<dyn DynamicResolver>>,
) -> PlaceholderEngine {
    let mut engine = PlaceholderEngine::new();
    let episodes_resolver = episodes
        .unwrap_or_else(|| crate::placeholder::EpisodesResolver::new(data_dir.into()).into_arc());
    engine.register_namespace("episodes", episodes_resolver);
    if let Some(host) = memory_host {
        engine.register_namespace(
            "memory",
            crate::placeholder::MemoryResolver::new(host).into_arc(),
        );
    }
    engine
}

#[tonic::async_trait]
impl Placeholder for PlaceholderService {
    async fn render(
        &self,
        request: Request<RenderRequest>,
    ) -> Result<Response<RenderResponse>, Status> {
        let req = request.into_inner();

        // Re-hydrate the engine context. The proto message allows empty
        // `model_name` to mean "none"; the Rust struct encodes that as
        // `Option::None`, so round-trip the sentinel.
        let mut ctx = PlaceholderCtx::new(
            req.ctx
                .as_ref()
                .map(|c| c.session_key.clone())
                .unwrap_or_default(),
        );
        if let Some(c) = req.ctx.as_ref() {
            if !c.model_name.is_empty() {
                ctx.model_name = Some(c.model_name.clone());
            }
            ctx.metadata = c.metadata.clone();
        }

        // Honour per-call max_depth override. 0 = use engine default
        // (matches the proto docstring).
        let engine = if req.max_depth == 0 {
            self.engine.clone()
        } else {
            Arc::new(self.engine.clone_with_max_depth(req.max_depth))
        };

        match engine.render(&req.template, &ctx).await {
            Ok(rendered) => {
                let unresolved = collect_unresolved(&rendered);
                Ok(Response::new(RenderResponse {
                    rendered,
                    unresolved_keys: unresolved,
                    error: String::new(),
                }))
            }
            Err(err) => Ok(Response::new(RenderResponse {
                rendered: String::new(),
                unresolved_keys: Vec::new(),
                error: encode_error(&err),
            })),
        }
    }
}

/// Encode a [`CorlinmanError::Parse`] that originated from a
/// `PlaceholderError` back into the stable wire form.
///
/// The engine funnels every variant through `CorlinmanError::Parse { what:
/// "placeholder", message }`, and the `Display` impl of `PlaceholderError`
/// produces deterministic prefixes ("placeholder cycle detected at key
/// '...'", "placeholder recursion depth ... exceeded", "resolver for '...'
/// failed: ..."). We match on those prefixes rather than coupling to the
/// `Display` string verbatim — the wire format is an explicit contract.
fn encode_error(err: &CorlinmanError) -> String {
    // The engine returns `CorlinmanError::Parse { what: "placeholder", message }`
    // whose Display is `"parse error (placeholder): <inner>"`. Strip that
    // wrapper so we're matching on `PlaceholderError::Display` directly.
    let raw = err.to_string();
    let inner = raw
        .strip_prefix("parse error (placeholder): ")
        .unwrap_or(&raw);

    // `PlaceholderError::Cycle` → `"placeholder cycle detected at key '<k>'"`
    if let Some(rest) = inner.strip_prefix("placeholder cycle detected at key '") {
        if let Some(key) = rest.strip_suffix('\'') {
            return format!("cycle:{key}");
        }
    }
    if inner.starts_with("placeholder recursion depth ") {
        return "depth_exceeded".into();
    }
    if let Some(rest) = inner.strip_prefix("resolver for '") {
        // rest = "<ns>' failed: <inner>"
        if let Some((_, tail)) = rest.split_once("' failed: ") {
            return format!("resolver:{tail}");
        }
    }
    // Unknown shape — surface verbatim so the Python side can still log
    // something actionable.
    format!("resolver:{inner}")
}

/// Harvest still-literal `{{…}}` tokens from a rendered template. The
/// engine preserves unknown tokens verbatim, so a post-render scan is
/// the cheapest way to surface them without modifying the engine.
fn collect_unresolved(rendered: &str) -> Vec<String> {
    if !rendered.contains("{{") {
        return Vec::new();
    }
    let mut out: Vec<String> = Vec::new();
    for cap in TOKEN_RE.captures_iter(rendered) {
        if let Some(m) = cap.get(1) {
            let body = m.as_str().trim();
            if body.is_empty() {
                continue; // `{{}}` / `{{ }}` are intentionally preserved
            }
            if !out.iter().any(|x| x == body) {
                out.push(body.to_string());
            }
        }
    }
    out
}

/// Bind a tonic `Server` onto `socket_path` and serve the `Placeholder`
/// service until `shutdown` resolves. Removes the socket file on exit so
/// subsequent boots can rebind cleanly.
///
/// Non-fatal: returns `Err` so `main.rs` can log-and-continue if binding
/// fails (e.g. permission denied on a read-only fs); the axum side keeps
/// serving `/v1/*` regardless.
pub async fn serve<F>(
    socket_path: impl AsRef<Path>,
    service: PlaceholderService,
    shutdown: F,
) -> anyhow::Result<()>
where
    F: std::future::Future<Output = ()> + Send + 'static,
{
    let path: PathBuf = socket_path.as_ref().to_path_buf();
    // Best-effort cleanup of a stale socket — a previous crash may have
    // left the file behind.
    let _ = tokio::fs::remove_file(&path).await;
    if let Some(parent) = path.parent() {
        tokio::fs::create_dir_all(parent).await.ok();
    }
    let listener = UnixListener::bind(&path)?;
    tracing::info!(socket = %path.display(), "placeholder gRPC bound");

    let incoming = UnixListenerStream::new(listener);
    let result = Server::builder()
        .add_service(PlaceholderServer::new(service))
        .serve_with_incoming_shutdown(incoming, shutdown)
        .await;

    let _ = tokio::fs::remove_file(&path).await;
    result.map_err(|e| anyhow::anyhow!("placeholder gRPC server exited: {e}"))
}

#[cfg(test)]
mod tests {
    use super::*;
    use corlinman_core::placeholder::PlaceholderEngine;
    use corlinman_proto::v1::PlaceholderCtx as PbCtx;

    fn service_with(engine: PlaceholderEngine) -> PlaceholderService {
        PlaceholderService::new(Arc::new(engine))
    }

    fn request(template: &str) -> Request<RenderRequest> {
        Request::new(RenderRequest {
            template: template.into(),
            ctx: Some(PbCtx {
                session_key: "s1".into(),
                model_name: "test-model".into(),
                metadata: Default::default(),
            }),
            max_depth: 0,
        })
    }

    fn request_with_max_depth(template: &str, max_depth: u32) -> Request<RenderRequest> {
        Request::new(RenderRequest {
            template: template.into(),
            ctx: Some(PbCtx {
                session_key: "s1".into(),
                model_name: "test-model".into(),
                metadata: Default::default(),
            }),
            max_depth,
        })
    }

    #[tokio::test]
    async fn static_hit_is_rendered() {
        let eng = PlaceholderEngine::new().with_static("date.today", "2026-04-22");
        let svc = service_with(eng);
        let resp = svc
            .render(request("today is {{date.today}}"))
            .await
            .unwrap()
            .into_inner();
        assert_eq!(resp.rendered, "today is 2026-04-22");
        assert!(resp.unresolved_keys.is_empty());
        assert!(resp.error.is_empty());
    }

    #[tokio::test]
    async fn unresolved_token_round_trips() {
        let svc = PlaceholderService::with_empty_engine();
        let resp = svc
            .render(request("hi {{var.user_name}} — {{session.id}}"))
            .await
            .unwrap()
            .into_inner();
        assert_eq!(resp.rendered, "hi {{var.user_name}} — {{session.id}}");
        assert_eq!(
            resp.unresolved_keys,
            vec!["var.user_name".to_string(), "session.id".to_string()]
        );
        assert!(resp.error.is_empty());
    }

    #[tokio::test]
    async fn empty_tokens_are_not_listed_as_unresolved() {
        let svc = PlaceholderService::with_empty_engine();
        let resp = svc
            .render(request("a {{}} b {{ }} c"))
            .await
            .unwrap()
            .into_inner();
        assert_eq!(resp.rendered, "a {{}} b {{ }} c");
        assert!(resp.unresolved_keys.is_empty());
    }

    #[tokio::test]
    async fn ctx_fields_round_trip_into_engine_context() {
        // Register a tiny dynamic resolver that echoes back a metadata value
        // so we can prove the Python-provided ctx reaches the engine.
        use async_trait::async_trait;
        use corlinman_core::placeholder::{DynamicResolver, PlaceholderError};

        struct Echo;
        #[async_trait]
        impl DynamicResolver for Echo {
            async fn resolve(
                &self,
                key: &str,
                ctx: &corlinman_core::placeholder::PlaceholderCtx,
            ) -> Result<String, PlaceholderError> {
                Ok(format!(
                    "{key}|{}|{}|{}",
                    ctx.session_key,
                    ctx.model_name.clone().unwrap_or_default(),
                    ctx.metadata.get("trace").cloned().unwrap_or_default(),
                ))
            }
        }

        let eng = PlaceholderEngine::new().with_dynamic("echo", Arc::new(Echo));
        let svc = service_with(eng);
        let mut meta = std::collections::HashMap::new();
        meta.insert("trace".to_string(), "t-42".to_string());
        let req = Request::new(RenderRequest {
            template: "{{echo.x}}".into(),
            ctx: Some(PbCtx {
                session_key: "sess-a".into(),
                model_name: "gpt".into(),
                metadata: meta,
            }),
            max_depth: 0,
        });
        let resp = svc.render(req).await.unwrap().into_inner();
        assert_eq!(resp.rendered, "x|sess-a|gpt|t-42");
        assert!(resp.error.is_empty());
    }

    #[tokio::test]
    async fn max_depth_override_preserves_registered_resolvers() {
        use async_trait::async_trait;
        use corlinman_core::placeholder::{DynamicResolver, PlaceholderError};

        struct Echo;
        #[async_trait]
        impl DynamicResolver for Echo {
            async fn resolve(
                &self,
                key: &str,
                _ctx: &corlinman_core::placeholder::PlaceholderCtx,
            ) -> Result<String, PlaceholderError> {
                Ok(format!("resolved:{key}"))
            }
        }

        let eng = PlaceholderEngine::new().with_dynamic("echo", Arc::new(Echo));
        let svc = service_with(eng);
        let resp = svc
            .render(request_with_max_depth("{{echo.memory}}", 2))
            .await
            .unwrap()
            .into_inner();

        assert_eq!(resp.rendered, "resolved:memory");
        assert!(resp.unresolved_keys.is_empty());
        assert!(resp.error.is_empty());
    }

    #[tokio::test]
    async fn production_engine_registers_episodes_and_memory_resolvers() {
        use async_trait::async_trait;
        use corlinman_core::placeholder::{DynamicResolver, PlaceholderError};
        use corlinman_memory_host::{MemoryDoc, MemoryHit, MemoryHost, MemoryQuery};

        struct StubEpisodes;
        #[async_trait]
        impl DynamicResolver for StubEpisodes {
            async fn resolve(
                &self,
                key: &str,
                _ctx: &corlinman_core::placeholder::PlaceholderCtx,
            ) -> Result<String, PlaceholderError> {
                Ok(format!("episodes:{key}"))
            }
        }

        struct StubMemory;
        #[async_trait]
        impl MemoryHost for StubMemory {
            fn name(&self) -> &str {
                "stub-memory"
            }

            async fn query(&self, req: MemoryQuery) -> anyhow::Result<Vec<MemoryHit>> {
                assert_eq!(req.text, "backend");
                assert_eq!(req.namespace.as_deref(), Some("agent-brain"));
                Ok(vec![MemoryHit {
                    id: "m1".into(),
                    content: "memory hit".into(),
                    score: 1.0,
                    source: "stub-memory".into(),
                    metadata: serde_json::Value::Null,
                }])
            }

            async fn upsert(&self, _doc: MemoryDoc) -> anyhow::Result<String> {
                anyhow::bail!("not used")
            }

            async fn delete(&self, _id: &str) -> anyhow::Result<()> {
                anyhow::bail!("not used")
            }
        }

        let engine = build_engine_with_episodes(
            PathBuf::from("."),
            Some(Arc::new(StubMemory)),
            Some(Arc::new(StubEpisodes)),
        );

        let out = engine
            .render(
                "{{episodes.recent}}\n{{memory.backend}}",
                &corlinman_core::placeholder::PlaceholderCtx::new("s"),
            )
            .await
            .unwrap();

        assert!(out.contains("episodes:recent"));
        assert!(out.contains("memory hit"));
        assert!(out.contains("stub-memory:m1"));
    }

    #[tokio::test]
    async fn cycle_error_is_encoded_with_key() {
        use async_trait::async_trait;
        use corlinman_core::placeholder::{DynamicResolver, PlaceholderError};

        struct SelfRef;
        #[async_trait]
        impl DynamicResolver for SelfRef {
            async fn resolve(
                &self,
                key: &str,
                _ctx: &corlinman_core::placeholder::PlaceholderCtx,
            ) -> Result<String, PlaceholderError> {
                // Every lookup returns a placeholder pointing to the same key,
                // triggering the engine's cycle guard on re-expansion.
                Ok(format!("{{{{loop.{key}}}}}"))
            }
        }

        let eng = PlaceholderEngine::new().with_dynamic("loop", Arc::new(SelfRef));
        let svc = service_with(eng);
        let resp = svc
            .render(request("{{loop.x}}"))
            .await
            .unwrap()
            .into_inner();
        assert!(resp.rendered.is_empty());
        assert_eq!(resp.error, "cycle:loop.x");
    }

    #[tokio::test]
    async fn resolver_error_is_encoded() {
        use async_trait::async_trait;
        use corlinman_core::placeholder::{DynamicResolver, PlaceholderError};

        struct Boom;
        #[async_trait]
        impl DynamicResolver for Boom {
            async fn resolve(
                &self,
                _key: &str,
                _ctx: &corlinman_core::placeholder::PlaceholderCtx,
            ) -> Result<String, PlaceholderError> {
                Err(PlaceholderError::Resolver {
                    namespace: "boom".into(),
                    message: "kaboom".into(),
                })
            }
        }

        let eng = PlaceholderEngine::new().with_dynamic("boom", Arc::new(Boom));
        let svc = service_with(eng);
        let resp = svc
            .render(request("{{boom.x}}"))
            .await
            .unwrap()
            .into_inner();
        assert!(resp.rendered.is_empty());
        assert_eq!(resp.error, "resolver:kaboom");
    }
}

File: python/packages/corlinman-server/src/corlinman_server/gateway/grpc/placeholder.py
"""gRPC wrapper over a Python ``PlaceholderEngine``.

Port of :rust:`corlinman_gateway::grpc::placeholder`. Direction: a
client (the Rust subsystem, the admin shell, the future Python
``context_assembler``) dials this service on the UDS path in
``$CORLINMAN_UDS_PATH`` (default ``/tmp/corlinman.sock``) and calls
``Render`` for every template it wants expanded before a provider call.

The :class:`PlaceholderEngine` Python sibling has not landed yet (it's
the W3 port of ``corlinman-core::placeholder``); we accept a structural
:class:`PlaceholderEngineLike` protocol so this module is testable today
and the eventual concrete engine drops in without touching this file.

Tokens with a namespace that has no resolver round-trip back unchanged
and are surfaced in ``RenderResponse.unresolved_keys`` for observability
— same contract as the Rust ``collect_unresolved`` post-render scan.

Error mapping preserves the enum shape of the Rust ``PlaceholderError``
so a single client library can dial either implementation:

==========================  =========================
engine error                ``error`` string
==========================  =========================
``CycleError(k)``           ``"cycle:<k>"``
``DepthExceededError(...)`` ``"depth_exceeded"``
``ResolverError(ns, msg)``  ``"resolver:<msg>"``
==========================  =========================
"""

from __future__ import annotations

import asyncio
import contextlib
import logging
import os
import re
from pathlib import Path
from typing import Any, Awaitable, Callable, Protocol, runtime_checkable

import grpc
from corlinman_grpc._generated.corlinman.v1 import (
    placeholder_pb2,
    placeholder_pb2_grpc,
)

__all__ = [
    "DEFAULT_RUST_SOCKET",
    "ENV_RUST_SOCKET",
    "PlaceholderCtx",
    "PlaceholderEngineLike",
    "PlaceholderError",
    "PlaceholderService",
    "collect_unresolved",
    "encode_error",
    "serve",
]


log = logging.getLogger(__name__)


# ─── Constants ────────────────────────────────────────────────────────

DEFAULT_RUST_SOCKET: str = "/tmp/corlinman.sock"
"""Default UDS path the gateway binds for Python→gateway traffic.

Kept separate from ``/tmp/corlinman-py.sock`` (the agent socket) so the
two sides can be restarted independently without stepping on each
other's socket file. Mirrors the Rust ``DEFAULT_RUST_SOCKET`` constant.
"""

ENV_RUST_SOCKET: str = "CORLINMAN_UDS_PATH"
"""Env var the Python ``PlaceholderClient`` honours, and the server
respects when set."""


# Mirrors the Rust ``TOKEN_RE`` lazy regex — same shape so the post-render
# unresolved-key scan finds the same tokens the engine would have tried
# to expand.
_TOKEN_RE: re.Pattern[str] = re.compile(r"\{\{([^{}]*?)\}\}")


# ─── Engine protocol (PlaceholderEngine port stub) ───────────────────


class PlaceholderCtx:
    """Render-time context handed to every resolver.

    Mirrors :rust:`corlinman_core::placeholder::PlaceholderCtx`. The
    actual Python ``PlaceholderEngine`` will own a richer version of
    this type; we keep a minimal shim here so the bridge can be
    constructed and tested without a hard dep on the (unported) engine.
    """

    __slots__ = ("session_key", "model_name", "metadata")

    def __init__(
        self,
        session_key: str,
        *,
        model_name: str | None = None,
        metadata: dict[str, str] | None = None,
    ) -> None:
        self.session_key = session_key
        self.model_name = model_name
        self.metadata: dict[str, str] = dict(metadata or {})


class PlaceholderError(Exception):
    """Base class for the three documented placeholder error shapes."""


class CycleError(PlaceholderError):
    """Cycle detected at ``key``."""

    def __init__(self, key: str) -> None:
        super().__init__(f"placeholder cycle detected at key '{key}'")
        self.key = key


class DepthExceededError(PlaceholderError):
    """Recursion depth limit reached."""

    def __init__(self, depth: int) -> None:
        super().__init__(f"placeholder recursion depth {depth} exceeded")
        self.depth = depth


class ResolverError(PlaceholderError):
    """Resolver raised for ``namespace``."""

    def __init__(self, namespace: str, message: str) -> None:
        super().__init__(f"resolver for '{namespace}' failed: {message}")
        self.namespace = namespace
        self.message = message


@runtime_checkable
class PlaceholderEngineLike(Protocol):
    """Structural surface the bridge needs.

    Mirrors the slice of the Rust ``PlaceholderEngine`` API used by the
    gRPC wrapper. Concrete impls will land alongside the
    ``PlaceholderEngine`` Python port; tests can wire in a fake.
    """

    async def render(self, template: str, ctx: PlaceholderCtx) -> str: ...

    def clone_with_max_depth(self, max_depth: int) -> PlaceholderEngineLike: ...


# ─── Service ──────────────────────────────────────────────────────────


class PlaceholderService(placeholder_pb2_grpc.PlaceholderServicer):
    """gRPC service shell.

    Wraps a shared :class:`PlaceholderEngineLike` so multiple concurrent
    ``Render`` RPCs share the same resolver registry. The engine is
    accepted as ``Optional`` so callers can stand up a no-resolver
    service for tests / boot-time bridges where every token round-trips
    back through ``unresolved_keys``.
    """

    def __init__(self, engine: PlaceholderEngineLike | None) -> None:
        self._engine = engine

    @classmethod
    def with_empty_engine(cls) -> PlaceholderService:
        """Convenience for tests + the equivalent of the Rust
        ``PlaceholderService::with_empty_engine``.

        Returns a service whose engine echoes every template back
        verbatim (i.e. no resolvers registered). Every ``{{ns.name}}``
        token is surfaced via ``unresolved_keys``.
        """
        return cls(_NullEngine())

    async def Render(  # noqa: N802 — gRPC casing
        self,
        request: placeholder_pb2.RenderRequest,
        context: grpc.aio.ServicerContext,
    ) -> placeholder_pb2.RenderResponse:
        # Re-hydrate the engine context. The proto message allows an
        # empty ``model_name`` to mean "none"; the Python ctx encodes
        # that as ``None`` so round-trip the sentinel.
        ctx_msg = request.ctx
        ctx = PlaceholderCtx(
            session_key=ctx_msg.session_key if ctx_msg is not None else "",
            model_name=(ctx_msg.model_name or None) if ctx_msg is not None else None,
            metadata=dict(ctx_msg.metadata) if ctx_msg is not None else None,
        )

        # Honour per-call ``max_depth`` override. 0 = use engine default
        # (matches the proto docstring + the Rust branch).
        engine = self._engine
        if engine is None:
            return placeholder_pb2.RenderResponse(
                rendered="",
                unresolved_keys=[],
                error="resolver:engine not configured",
            )
        if request.max_depth != 0:
            engine = engine.clone_with_max_depth(int(request.max_depth))

        try:
            rendered = await engine.render(request.template, ctx)
        except PlaceholderError as err:
            return placeholder_pb2.RenderResponse(
                rendered="",
                unresolved_keys=[],
                error=encode_error(err),
            )
        except Exception as err:  # noqa: BLE001 — surface as resolver error
            # Unknown shapes — surface verbatim so the client can still
            # log something actionable. Mirrors the Rust ``encode_error``
            # fallback branch.
            return placeholder_pb2.RenderResponse(
                rendered="",
                unresolved_keys=[],
                error=f"resolver:{err}",
            )

        unresolved = collect_unresolved(rendered)
        return placeholder_pb2.RenderResponse(
            rendered=rendered,
            unresolved_keys=unresolved,
            error="",
        )


class _NullEngine:
    """Engine sibling of :rust:`PlaceholderEngine::new()` with zero
    resolvers — every template echoes back verbatim so the post-render
    scan surfaces every token as unresolved."""

    async def render(self, template: str, ctx: PlaceholderCtx) -> str:
        return template

    def clone_with_max_depth(self, max_depth: int) -> _NullEngine:
        return self


# ─── Helpers (pure) ───────────────────────────────────────────────────


def encode_error(err: Exception) -> str:
    """Encode a Python placeholder error back into the stable wire form.

    Mirrors :rust:`encode_error` byte-for-byte:

    * :class:`CycleError`        → ``"cycle:<k>"``
    * :class:`DepthExceededError`→ ``"depth_exceeded"``
    * :class:`ResolverError`     → ``"resolver:<msg>"``
    * unknown / generic          → ``"resolver:<str(err)>"``
    """
    if isinstance(err, CycleError):
        return f"cycle:{err.key}"
    if isinstance(err, DepthExceededError):
        return "depth_exceeded"
    if isinstance(err, ResolverError):
        return f"resolver:{err.message}"
    # Tolerate "wrapped" errors coming up through a future
    # ``CorlinmanError::Parse`` lookalike: match the prefixes the Rust
    # encoder strips before classifying.
    raw = str(err)
    inner = raw.removeprefix("parse error (placeholder): ")

    if inner.startswith("placeholder cycle detected at key '") and inner.endswith("'"):
        key = inner[len("placeholder cycle detected at key '") : -1]
        return f"cycle:{key}"
    if inner.startswith("placeholder recursion depth "):
        return "depth_exceeded"
    if inner.startswith("resolver for '"):
        # "resolver for '<ns>' failed: <inner>"
        rest = inner[len("resolver for '") :]
        marker = "' failed: "
        if marker in rest:
            _, tail = rest.split(marker, 1)
            return f"resolver:{tail}"

    return f"resolver:{inner}"


def collect_unresolved(rendered: str) -> list[str]:
    """Harvest still-literal ``{{…}}`` tokens from a rendered template.

    The engine preserves unknown tokens verbatim, so a post-render scan
    is the cheapest way to surface them without modifying the engine.
    Mirrors :rust:`collect_unresolved` 1:1, including the
    empty-body skip (``{{}}`` / ``{{ }}`` are intentionally preserved
    so callers can use them as literal markup).
    """
    if "{{" not in rendered:
        return []
    out: list[str] = []
    for match in _TOKEN_RE.finditer(rendered):
        body = match.group(1).strip()
        if not body:
            continue
        if body not in out:
            out.append(body)
    return out


# ─── Server helper ────────────────────────────────────────────────────


async def serve(
    socket_path: str | os.PathLike[str],
    service: PlaceholderService,
    shutdown: asyncio.Event | Awaitable[None],
) -> None:
    """Bind a ``grpc.aio`` server onto ``socket_path`` and serve the
    ``Placeholder`` service until ``shutdown`` fires.

    Removes the socket file on exit so subsequent boots can rebind
    cleanly. Mirrors :rust:`serve` — the call is non-fatal in spirit:
    callers wrap it in a task and log-and-continue if binding fails
    (e.g. permission denied on a read-only fs).

    ``shutdown`` may be either an :class:`asyncio.Event` (set when ready
    to stop) or any awaitable that resolves when the server should
    shut down. Mirrors the Rust ``F: Future<Output = ()>`` bound.
    """
    path = Path(os.fspath(socket_path))

    # Best-effort cleanup of a stale socket — a previous crash may have
    # left the file behind. Matches the Rust cleanup-before-bind dance.
    with contextlib.suppress(FileNotFoundError, OSError):
        path.unlink()
    path.parent.mkdir(parents=True, exist_ok=True)

    server = grpc.aio.server()
    placeholder_pb2_grpc.add_PlaceholderServicer_to_server(service, server)
    # gRPC supports ``unix:`` URIs for UDS listeners.
    server.add_insecure_port(f"unix:{path}")
    await server.start()
    log.info("placeholder gRPC bound socket=%s", path)

    try:
        if isinstance(shutdown, asyncio.Event):
            await shutdown.wait()
        else:
            await shutdown
    finally:
        # Mirror the Rust ``serve_with_incoming_shutdown`` cleanup: try
        # a graceful stop first, then unlink the socket file.
        await server.stop(grace=1.0)
        with contextlib.suppress(FileNotFoundError, OSError):
            path.unlink()


# Re-export for typing convenience (matches Rust ``pub use`` pattern).
_unused_typing: tuple[Any, ...] = (
    Callable,
)  # keep imports flake-clean across linters that strip unused.
